css-highlight-api-1/Overview.bs

```diff
@@ -225,41 +214,45 @@ Registering Custom Highlights</h3>
 
 	[Exposed=Window]
 	interface HighlightRegistry {
-		setlike<Highlight>;
+		maplike<DOMString, Highlight>;
 	};
 	</xmp>
 
 	<div algorithm="to register a custom highlight">
 		To [=register=] a [=custom highlight=],
-		invoke the {{HighlightRegistry/add()}} method of the [=highlight registry=]
-		with the [=custom highlight=] as the argument.
-
-		When invoked,
-		the <dfn method for=HighlightRegistry>add(Highlight value)</dfn> method must run these steps:
-
-		1. If there is already a [=set entry=] with the same {{Highlight/name}} as the {{Highlight/name}} of {{value}},
-			then [=throw=] an {{"OperationError"}}.
-		3. Let |valueArg| be the result of [=converted to an ECMAScript value|converting=] {{value}} to an ECMAScript value.
-		4. Let |result| be the result of running [[webidl#es-add-delete|the steps for a built-in setlike add function]],
-			with the [=context object=] as the <code>this</code> value
-			and with |valueArg| as the argument.
-		5. Return |result|.
-	</div>
+		invoke the <code>set</code> method of the [=highlight registry=]
+		which will run [[webidl#es-map-set|the steps for a built-in maplike set function]],
+		with the [=context object=] as the <code>this</code> value,
+		the passed-in [=custom highlight name=] as |keyArg|,
+		and the passed-in highlight as |valueArg|.
+
+		The <dfn>custom highlight name</dfn> assigned to a [=custom highlight=] when it is [=registered=]
+		is used to identify the highlight during styling (see [[#styling-highlights]]).
+
+	Note: It is possible to [=register=] a [=custom highlight=] with any [=custom highlight name=]
+	that is a valid [[webidl#idl-DOMString|DOMString]], but the name must also be a valid CSS
+	<<ident-token>> in order to be used to style the highlight.
```

Comment:
This seems like a footgun, why not restrict these to valid `<ident-token>`s so that the error can be caught at registration time? Is there any other purpose for these highlights other than being styled in CSS?

Reply:
More importantly, this seems incorrect; you can *write* an <<ident-token>> whose value is *any* arbitrary string, it just might require appropriate use of escapes.  So this Note should be rewritten to instead be a *recommendation* for authors to provide names that are CSS idents, so that it's easier to use them in CSS, rather than implying they can't be referred to at all from CSS.

Reply:
@LeaVerou, @tabatkins: What I had in mind was that this should work similar to id on elements. For example, authors can assign ids to elements that are not ident tokens - ex. `div.setAttribute('id', '+')` - but authors can still target such elements from CSS using escapes - ex. `#\+ { color: red; }`. Similarly, the highlight OM APIs do not need to prevent authors from registering non-ident token names, and the following example should work (ie. 'b' should be red) :

```
<html>
<head>
    <style>
      div::highlight(\+) {
        color: red;
      }
    </style>
</head>
<body>
    <div>abd</div>
    <script>
      let div = document.querySelector('div');
      let textNode = div.firstChild;
      let range = new StaticRange({startContainer: textNode, startOffset: 1, endContainer: textNode, endOffset: 2});
      let highlight = new Highlight(range);
      CSS.highlights.set('+', highlight);
    </script>
</body>
</html>
```
Do you agree with this model, or would you like to see something different?

The rationale behind this note was that I wasn't sure if all character values accepted by DOMString could be represented as an ident token. I couldn't come up with an example of such a case, but it also wasn't clear to me from reading the CSS identifier and DOMString specs that there wasn't a case.
Thinking about this some more though, if there were a case like this then it would affect id and other similar selectors in the same way, and I don't see any note about this in the CSS syntax spec. I also don't see anything about this in the CSS.escape spec either. So I'm starting to think that there really may not be such a case (@tabatkins, that's what you were saying too, right?). If so, I think we can just remove the note.

@LeaVerou, just to respond to your second question as well, I assume you were referring to "highlight names" when you said "highlights"? It's true that highlight names are only used for styling via CSS, but I think it's still better to match the prior art we have here. (The highlight objects themselves are also used for setting priority, and they will also likely become event targets when we work on highlight events in the future.)

Reply:
Since you can always write an ident *solely* as a list of escapes (and you can encode any codepoint in escapes), it's always possible, if sometimes unwieldy, to encode *almost any* DOMString as a CSS ident. (The only exception is that idents can't contain U+0000 NULL, or any surrogate code points. But if you're putting that in your string you've placed your own foot entirely in the barrel of the gun, and it's not our fault you get shot.)

The general policy, from other CSS APIs where we compare or set a CSS ident with a JS string, is to just use the string directly, with no CSS parsing done on it. See <https://drafts.csswg.org/css-conditional/#dom-css-supports>, for example.

So yeah, let authors do what they want here, don't try to limit them. Leave in a note that *warns against* making strings that aren't CSS identifiers (linking `[=CSS/identifier=]`), but no more.

Reply:
Thanks for the information @tabatkins . Makes sense to leave a warning here then. I've made that update. PTAL.

Reply:
@tabatkins Not sure if you wanted to look at this again or not. I'm going to go ahead and merge this. Feel free to let me know if you had any additional feedback here, I can address it with a follow up change.